### ISSUE TITLE
HL-ZASM Proposal: Disable NoUnreferencedLeaves by default

### ISSUE THREAD

**Issue by R2ZER0** [wiremod/wire#494]
I propose that the HL-ZASM pragma setting NoUnreferencedLeaves be disabled by default.

For those who aren't sure what this does exactly, is it basically doesn't compile functions in your program than are never used, to save space. E.g., the example program:

``` c
main();

void main() {
    float i = 3;
    i += 5;
    out 1, i;
}

void foo() {
    out 2, 53;
}
```

In the resulting binary, the code for the funcion `foo()` isn't present at all, as it is never used.

On the surface this seems like a good idea, and in principle it is, but i think it should require the programmer to choose to turn it on. This is in line with other compilers, where the programmer chooses to enable optimisations, and i would definately consider this to be an optimisation. Also, the compiler should by default do exactly what you expect it to, and unless you happen to know the HL-ZASM compiler fairly well, you probably wouldn't be aware this is happenning.

A major argument for disabling by default however, is that it's and endless source of mysterious bugs. For instance, trying to compile this program will fail to compile:

``` c
main();

void main() {
    // do something, but NOT using foo()
}

void foo() {
    bar("something");
}

void bar() {
    // ...
}
```

With error `[generic 10:20] Must be constant expression: "something"`.
Now unless you already knew why this happened, you would probably think along the lines of "WTF does that mean?!?". This, i think, is a major contributor to HL-ZASM/the CPU getting it's reputation for bugginess.

Also it seems to cause odd bugs here and there, for instance compiling the GPU's 3d cube example vanilla will fail, but adding the magic `#pragma set NoUnreferencedLeaves false` to the top of the gl driver library file will suddenly make everything work again. Nowadays i automatically add the line to any program i write, simply because it solves so many bugs.

It must cause endless headaches for beginning and experianced ZCPU programmers alike, and i've seen several forum discussions of seemingly impossible to track down bugs caused by exactly this.

So in conclusion, disabling this by default would both make the compiler more intuitive, and get rid of vast swathes of bugs at once, while still allowing those who want to use it to... so i think we should!

Of course it would be nice if we could fix the bugs in the feature, but that is a whole nother investigation.
